### PR TITLE
First release of MHC Module

### PR DESCRIPTION
**Pull request by MrmcSteve**
MHC Module has configuration of IP address, port, username and password for basic authentication for https API.  Added actions for recalling presets, moves and stopping.

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -27,23 +30,105 @@ class ModuleInstance extends InstanceBase {
...
+					url,
+					{
+						username:self.config.username,
+						password:self.config.password,
+						https: { rejectUnauthorized: false }
+					}
+				)
+				console.log("Result from REST:" + str);
+			} catch (err) {
+				throw new Error(`Action failed err: ${err}`)
```
Comment (dnmeid) [inline main.js:54]: Suggest to reference which action failed and do a more user oriented error handling, e.g. setting status and logging error message.

File: main.js [hunk 2]
```diff
@@ -27,23 +30,105 @@ class ModuleInstance extends InstanceBase {
...
 	// Return config fields for web config
 	getConfigFields() {
 		return [
 			{
 				type: 'textinput',
-				id: 'host',
+				id: 'httpHost',
 				label: 'Target IP',
 				width: 8,
 				regex: Regex.IP,
```
Comment (dnmeid) [inline main.js:110]: Why limit to IP if you can use also hostnames?

File: package.json [hunk 3]
```diff
@@ -12,7 +12,8 @@
 		"url": "git+https://github.com/bitfocus/companion-module-your-module-name.git"
```
Comment (dnmeid) [inline package.json:12]: fill name of your module

File: variables.js [hunk 4]
```diff
@@ -1,4 +1,8 @@
 module.exports = async function (self) {
+
+	const x =self.getRobotList(self)
+	console.log(x)
```
Comment (dnmeid) [inline variables.js:4]: remove

File: variables.js [hunk 5]
```diff
@@ -1,4 +1,8 @@
...
+
 	self.setVariableDefinitions([
```
Comment (dnmeid) [inline variables.js:6]: implement with your variables

File: main.js [hunk 6]
```diff
@@ -27,23 +30,105 @@ class ModuleInstance extends InstanceBase {
...
+		try {
+			const response = await got.get(
+				`https://${self.config.httpHost}:${self.config.httpPort}/v1/status`,
+				{
+					username:self.config.username,
+					password:self.config.password,
+					https: { rejectUnauthorized: false }
+				}
+			).json()
+			console.log("Result from REST:" + response);
```
Comment (dnmeid) [inline main.js:96]: remove

File: main.js [hunk 7]
```diff
@@ -27,23 +30,105 @@ class ModuleInstance extends InstanceBase {
...
+		} catch (err) {
+			return []
+		}
+	}
```
Comment (dnmeid) [inline main.js:100]: function tries to get response but doesn't return it

File: main.js [hunk 8]
```diff
@@ -27,23 +30,105 @@ class ModuleInstance extends InstanceBase {
...
+						password:self.config.password,
+						https: { rejectUnauthorized: false }
+					}
+				)
+				console.log("Result from REST:" + str);
+			} catch (err) {
+				throw new Error(`Action failed err: ${err}`)
+			}
+		}
+	}
```
Comment (dnmeid) [inline main.js:83]: function tries to get response but doesn't return it

File: main.js [hunk 9]
```diff
@@ -27,23 +30,105 @@ class ModuleInstance extends InstanceBase {
 		this.config = config
 	}
 
+	async sendCam(self, str) {
+		const { default: got } = await import('got');
```
Comment (dnmeid) [inline main.js:34]: why importing got in every function locally?

File: main.js [hunk 10]
```diff
@@ -27,23 +30,105 @@ class ModuleInstance extends InstanceBase {
...
+		if (str) {
+
+			const url = `https://${self.config.httpHost}:${self.config.httpPort}${str}`
```
Comment (dnmeid) [inline main.js:37]: this is super generic, why call the function postCam and then use only a generic url. Suggest to write only one function with a generic url and also a generic name and call that from more specialized functions (dry principle)

File: main.js [hunk 11]
```diff
@@ -27,23 +30,105 @@ class ModuleInstance extends InstanceBase {
 		this.config = config
 	}
 
+	async sendCam(self, str) {
```
Comment (dnmeid) [inline main.js:33]: no need to pass self to a method of the class here

File: main.js [hunk 12]
```diff
@@ -27,23 +30,105 @@ class ModuleInstance extends InstanceBase {
...
+			try {
+				const response = await got.put(
+					url,
+					{
+						username:self.config.username,
+						password:self.config.password,
+						https: { rejectUnauthorized: false }
+					}
+				)
+				console.log("Result from REST:" + str);
```
Comment (dnmeid) [inline main.js:52]: remove

File: main.js [hunk 13]
```diff
@@ -27,23 +30,105 @@ class ModuleInstance extends InstanceBase {
...
+			} catch (err) {
+				throw new Error(`Action failed err: ${err}`)
+			}
+		}
+	}
+
+	async postCam(self, str) {
```
Comment (dnmeid) [inline main.js:59]: only difference to sendCam is the http method. DRY

File: companion/manifest.json [hunk 14]
```diff
@@ -20,7 +20,7 @@
 		"entrypoint": "../main.js"
 	},
 	"legacyIds": [],
-	"manufacturer": "Your company",
-	"products": ["Your product"],
+	"manufacturer": "Mark Roberts Motion Control",
+	"products": ["MHC"],
 	"keywords": []
```
Comment (dnmeid) [inline companion/manifest.json:25]: Suggest adding some keywords

**Review by dnmeid — CHANGES_REQUESTED** [review]
You probably don't want to go with example feedback.
Please request review only when you are finished, this looks untested.

File: main.js [hunk 15]
```diff
@@ -60,4 +111,4 @@ class ModuleInstance extends InstanceBase {
 	}
```
Comment (dnmeid) [inline main.js:111]: UpdateFeedbacks and UpdateVariableDefinitions now is undefined.

Comment (mihindupaul) [reply main.js:111]: We have no plan to support feedbacks or variables for the initial version, is it valid to initialize feedbacks and variables with empty structures ?

Comment (dnmeid) [reply main.js:111]: You don't need to call `updateFeedbacks` or `updateVariableDefinitions` at all and in deed you are not calling these functions. But what you do is to declare them and within these function definitions you are calling some formerly importet functions from other modules/files which are now not defined any more.
Just remove the unused function definitions.
But I can tell you that feedback and presets are greatly improving user experience and so far your module is very basic.

Comment (mihindupaul) [reply main.js]: Thanks for the heads-up. variables and feedback modules are now temporarily disabled until  API support for feedbacks and variables are improved.

File: package.json [hunk 16]
```diff
@@ -1,5 +1,5 @@
 {
-	"name": "your-module-name",
+	"name": "mrmoco-mhc",
 	"version": "0.1.0",
```
Comment (dnmeid) [inline package.json:3]: If this is going to be the first release, please change version here to 1.0.0 and leave in manifest.json at 0.0.0

**Review by dnmeid — CHANGES_REQUESTED** [review]

**Review by dnmeid — APPROVED** [review]

**Pull request merged**